docs/aep/AEP-2-table-cleanup.md

```diff
@@ -0,0 +1,186 @@
+* Allows using tasks trigger rules which gives control of running cleanup even when certain task fails
+
+Cons:
+* If users add new tasks to the DAG, they need to remember to add it to the list
+* It might find all tables and delete them before they necessarily SHOULD be deleted if the downstream tasks are not added to the dependencies list
+
+### Option 4: a function with no parameters
+
+```python
+with dag:
+    my_homes_table = aql.load_file(
+        input_file=File(path=f"{s3_bucket}/homes.csv"),
+        output_table=Table(
+            conn_id="postgres_conn",
+        ),
+    )
+    sample_table = sample_create_table(my_homes_table)
+    df_res = my_df_func(sample_table)
+    [my_homes_table, sample_table, df_res] >> aql.cleanup()
+```
+
+Pros:
+* Would automatically find all tables
+* Easy to debug, as tasks logs are accessible from the webserver as opposed to callbacks
+* Allows using tasks trigger rules which gives control of running cleanup even when certain task fails
+
+Cons:
+* This ruins the feeling of the astro python SDK seeming “new”. We are exiting the functional paradigm and are now back in the land of bitshifting tasks.
+* If users add new tasks to the DAG, they need to remember to add it to the list
+* It might find all tables and delete them before they necessarily SHOULD be deleted if the downstream tasks are not added to the dependencies list
+
+### Option 5: A listener
+
+We looked into using the new airflow listener feature for this feature, but unfortunately found it was not compatible. Even if it was, it would require
+asking users to install a custom plugin to their airflow instance which would add startup overhead.
+
+### Option 6: An asynchronous task
+
+One major issue with all task-based ideas is that we end up in a situation where the DAG becomes much harder to read (as there is a final task that is downstream of all tasks. This can create a lot of clutter
+
+Where a DAG before might look like this:
+
+![](img/AEP-2/example-dag.png)
+
+The addition of the extra step adds a lot of complexity to the DAG shape:
+
+![](img/AEP-2/example-with-downstream-cleanup.png)
+
+Ideally we would like to have the transparency of a task, but without the extra logic or DAG structure complexity.
+
+We believe that we can accomplish this using an asynchronous operator.
+
+Using an asynchronous operator, we can create a task that does not NEED to run downstream of every other task. Instead this task can run in parallel, and only actually perform cleanups once every other task in the DAG has been accounted for (either passing, failing, or being skipped).
+
+![](img/AEP-2/example-with-async-task.png)
+
+This structure will allow the DAG structure to remain relatively unchanged, keep additional code to a minimum for the DAG author AND allow users the ability to control if they only want certain tables deleted (e.g. if you give it a list of tasks it will only delete for those tasks, otherwise it will wait for all tasks to complete).
```

Comment:
@dimberman How will the _cleanup task know what tables to delete and keep? will it be based on naming convention(_tmp<random_string>) or will it delete all tables?